File: tile.cpp
#include <iostream>
#include <string>
#include <iterator> 
#include <vector>
#include <stdio.h>      
#include <stdlib.h>    
#include "engimon.hpp"
#include "skill.hpp"
#include "tile.hpp"
using namespace std;

#define start_seed 1357

int tile::random_seed = start_seed;
engimon tile::universal_nullEngimon;
tile::tile()
{
	this->x = -1;
	this->y = -1;
	engimon nullEngimon;
	this->wild_engimon = nullEngimon;
	this->symbol = '-';
	this->tile_type = "grassland";
	this->tilePass = false;
}
tile::tile(int x, int y, engimon e, string type)
{
	this->x = x;
	this->y = y;
	this->wild_engimon = e;
	this->tile_type = type; //Hanya berupa grassland atau sea

	if (type == "sea")
	{
		symbol = 'o';
	}
	else
	{
		symbol = '-';
	}

	this->tilePass = false;

}
tile::~tile(){}

bool tile::isTileValid(engimon e)
{
	if ( ((e.getElmt1() == "Fire" || e.getElmt1() == "Ground" || e.getElmt1() == "Electric")
	|| (e.getElmt2() == "Fire" || e.getElmt2() == "Ground" || e.getElmt2() == "Electric"))
	&& this->tile_type == "grassland")
	{
		return true;
	}
	else if ( ((e.getElmt1() == "Water" || e.getElmt1() == "Ice")
	|| (e.getElmt2() == "Water" || e.getElmt2() == "Ice"))
	&& this->tile_type == "sea")
	{
		return true;
	}
	else
	{
		return false;
	}
}

bool tile::haveWildEngimon()
{
	if (this->wild_engimon.isNull())
	{
		return false;
	}
	else
	{
		return true;
	}
}

void tile::passed()
{
	this->tilePass = true;
}
void tile::resetPass()
{
	this->tilePass = false;
}
bool tile::isPass()
{
	return tilePass;
}
void tile::spawn()
{
	int species_count = engimon::all_species.size();

	/*Random module*/
	srand(tile::random_seed);
	tile::random_seed++;
	int randomNumber;
	randomNumber = rand() %  species_count;
	srand(tile::random_seed);
	tile::random_seed++;
	int randomLevel = rand() % 40 + 1;
	srand(tile::random_seed);
	tile::random_seed++;

	/*Find a species*/

	auto idx = engimon::all_species.begin();
	for (auto i = engimon::all_species.begin(); i != engimon::all_species.end() && randomNumber > 0; ++i)
	{
		randomNumber--;
		idx = i;
	}

	/*Pembuatan Engimon*/
	skill nullSkill;
	//engimon potentialEngimon(get<0>(*idx), &universal_nullEngimon, &universal_nullEngimon, get<0>(*idx), randomLevel, 0);


	wildEngimon potentialEngimon(get<0>(*idx), randomLevel);
		//get<0>(*idx) == nama species engimon

	while(!this->isTileValid(potentialEngimon))
	{
		randomNumber = rand() %  species_count;
		srand(tile::random_seed);
		tile::random_seed++;

		randomLevel = rand() % 30 + 1;
		srand(tile::random_seed);
		tile::random_seed++;

		for (auto i = engimon::all_species.begin(); i != engimon::all_species.end() && randomNumber > 0; ++i)
		{
			randomNumber--;
			idx = i;
		}

		//potentialEngimon = engimon(get<0>(*idx), &universal_nullEngimon, &universal_nullEngimon, get<0>(*idx), randomLevel, 0);
		potentialEngimon = wildEngimon(get<0>(*idx), randomLevel);
	}

	this->setEngimon(potentialEngimon);
	//cout << potentialEngimon.getName() <<" berhasil spawn di "<< x << ", " << y << endl;
}

void tile::printTile()
{
	cout << this->symbol;
}

void tile::setEngimon(engimon e)
{
	this->wild_engimon = e;
	this->updateCharacter();
}

engimon& tile::getEngimon()
{
	return this->wild_engimon;
}

void tile::setSymbol(char c)
{
	this->symbol = c;
}


void tile::playerIsHere()
{
	isPlayer = true;
	this->setSymbol('P');
}
void tile::playerIsNotHere()
{
	isPlayer = false;
	this->updateCharacter();
}
bool tile::isPlayerHere()
{
	return isPlayer;
}

void tile::activeEngimonIsHere()
{
	isActivePokemon = true;
	this->setSymbol('X');
}
void tile::activeEngimonIsNotHere()
{
	isActivePokemon = false;
	this->updateCharacter();
}
bool tile::isactiveEngimonHere()
{
	return isActivePokemon;
}

void tile::updateCharacter()
{
	if (this->wild_engimon.isNull())
	{
		if (this->tile_type == "sea")
		{
			this->symbol = 'o';
		}
		else
		{
			this->symbol = '-';
		}
	}
	else if (this->wild_engimon.getElmt1() == "Fire" && this->wild_engimon.getElmt2() == "Electric")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'L';
		}
		else
		{
			this->symbol = 'l';
		}
		
	}
	else if (this->wild_engimon.getElmt1() == "Water" && this->wild_engimon.getElmt2() == "Ice")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'S';
		}
		else
		{
			this->symbol = 's';
		}
		
	}

	else if (this->wild_engimon.getElmt1() == "Water" && this->wild_engimon.getElmt2() == "Ground")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'N';
		}
		else
		{
			this->symbol = 'n';
		}
		
	}

	else if (this->wild_engimon.getElmt1() == "Fire")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'F';
		}
		else
		{
			this->symbol = 'f';
		}
		
	}
	else if (this->wild_engimon.getElmt1() == "Water")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'W';
		}
		else
		{
			this->symbol = 'w';
		}
		
	}
	else if (this->wild_engimon.getElmt1() == "Electric")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'E';
		}
		else
		{
			this->symbol = 'e';
		}
	}

	else if (this->wild_engimon.getElmt1() == "Ground")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'G';
		}
		else
		{
			this->symbol = 'g';
		}
	}

	else if (this->wild_engimon.getElmt1() == "Ice")
	{
		if (this->wild_engimon.getLevel() > HGIH_LEVEL_CAP)
		{
			this->symbol = 'I';
		}
		else
		{
			this->symbol = 'i';
		}
	}
}

File: Java/tile.java
public class tile {

    protected int x;
    protected int y;

    private engimon tileEngimon;
    private char symbol;
    private String tile_type; //Hanya berupa mountains, tundra,grassland atau sea

    private Boolean tilePass;
    private Boolean isPlayer  = false;
    private Boolean isActivePokemon  = false;

    public tileListener listener;

    public tile(int x, int y, String type)
    {
        this.x = x;
        this.y = y;
        this.tile_type = type;
        if (tile_type.equals("mountains") )
        {
            this.symbol = ('^');
        }
        else if (tile_type.equals("sea"))
        {
            this.symbol = ('o');
        }
        else if (tile_type.equals("grassland"))
        {
            this.symbol = ('-');
        }
        else if (tile_type.equals("tundra"))
        {
            this.symbol = ('.');
        }
        this.tilePass = false;
        this.tileEngimon = null;
    }

    public tile(int x, int y, wild_engimon e, String type)
    {
        this(x,y,type);
        this.tileEngimon = e;
    }

    public Boolean isTileValid(engimon e)
    {
        if (tile_type.equals("mountains") && ("Fire".equals(e.getElmt1()) || "Fire".equals(e.getElmt2())))
        {
            return true;
        }
        else if (tile_type.equals("sea") && ("Water".equals(e.getElmt1()) || "Water".equals(e.getElmt2())))
        {
            return true;
        }
        else if ((tile_type.equals("grassland") && ("Ground".equals(e.getElmt1()) || "Ground".equals(e.getElmt2()))) ||
                (tile_type.equals("grassland") && ("Electric".equals(e.getElmt1()) || "Electric".equals(e.getElmt2()))))
        {
            return true;
        }
        else if (tile_type.equals("tundra") && ("Ice".equals(e.getElmt1()) || "Ice".equals(e.getElmt2())))
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    public Boolean haveWildEngimon()
    {
        return (this.tileEngimon != null);
    }

    public void passed()
    {
        this.tilePass = true;
    }
    public void resetPass()
    {
        this.tilePass = false;
    }
    public Boolean isPass()
    {
        return this.tilePass;
    }
    public void spawn()
    {
        wild_engimon we;
        try
        {
            we = new wild_engimon(this.tile_type);
            this.tileEngimon = we;
        }
        catch (Exception e)
        {
            e.printStackTrace();
            System.out.println(e.getMessage());
            this.tileEngimon = null;
        }



    }
    public void printTile()
    {
        System.out.print(symbol);
    }

    public void setEngimon(engimon e)
    {
        this.tileEngimon = e;
    }
    public engimon getEngimon()
    {
        return this.tileEngimon;
    }

    public String getTile_type() {
        return tile_type;
    }

    public void setListener(tileListener listener) {
        this.listener = listener;
    }

    public void setSymbol(char symbol)

    {
        this.symbol = symbol;
        if (this.listener != null)
        {
            listener.tileChangedCharacter(this.symbol, this.tileEngimon);
        }
    }

    public char getSymbol() {
        return symbol;
    }

    public void playerIsHere()
    {
        this.isPlayer = true;
        this.setSymbol('P');
    }
    public void playerIsNotHere()
    {
        this.isPlayer = false;
        this.updateCharacter();
    }
    public Boolean isPlayerHere()
    {
        return this.isPlayer;
    }

    public void activeEngimonIsHere()
    {
        this.isActivePokemon = true;
        this.setSymbol('X');
    }
    public void activeEngimonIsNotHere()
    {
        this.isActivePokemon = false;
        this.updateCharacter();
    }
    public Boolean isactiveEngimonHere()
    {
        return this.isActivePokemon;
    }

    public void updateCharacter()
    {
        if (tile_type.equals("mountains") && this.tileEngimon == null)
        {
            this.setSymbol('^');
        }
        else if (tile_type.equals("sea") && this.tileEngimon == null)
        {
            this.setSymbol('o');
        }
        else if (tile_type.equals("grassland") && this.tileEngimon == null)
        {
            this.setSymbol('-');
        }
        else if (tile_type.equals("tundra") && this.tileEngimon == null)
        {
            this.setSymbol('.');
        }
        else if (("Fire".equals(this.tileEngimon.getElmt1()) && "Electric".equals(this.tileEngimon.getElmt2())) ||
                ("Electric".equals(this.tileEngimon.getElmt1()) && "Fire".equals(this.tileEngimon.getElmt2())))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('L');
            }
            else
            {
                this.setSymbol('l');
            }

        }
        else if (("Water".equals(this.tileEngimon.getElmt1()) && "Ice".equals(this.tileEngimon.getElmt2())) ||
                ("Ice".equals(this.tileEngimon.getElmt1()) && "Water".equals(this.tileEngimon.getElmt2())))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('S');
            }
            else
            {
                this.setSymbol('s');
            }

        }

        else if (("Water".equals(this.tileEngimon.getElmt1()) && "Ground".equals(this.tileEngimon.getElmt2())) ||
                ("Ground".equals(this.tileEngimon.getElmt1()) && "Water".equals(this.tileEngimon.getElmt2())))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('N');
            }
            else
            {
                this.setSymbol('n');
            }

        }

        else if ("Fire".equals( this.tileEngimon.getElmt1()))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('F');
            }
            else
            {
                this.setSymbol('f');
            }

        }
        else if ("Water".equals( this.tileEngimon.getElmt1()))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('W');
            }
            else
            {
                this.setSymbol('w');
            }

        }
        else if ("Electric".equals( this.tileEngimon.getElmt1()))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('E');
            }
            else
            {
                this.setSymbol('e');
            }
        }

        else if ("Ground".equals( this.tileEngimon.getElmt1()))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('G');
            }
            else
            {
                this.setSymbol('g');
            }
        }

        else if ("Ice".equals( this.tileEngimon.getElmt1()))
        {
            if (this.tileEngimon.getLevel() > wild_engimon.playerHighestPokemon)
            {
                this.setSymbol('I');
            }
            else
            {
                this.setSymbol('i');
            }
        }
        else
        {
            this.setSymbol('?');
        }
    }

}
